### PR TITLE
Add Iterator Design Pattern

### PR DESCRIPTION
**Pull request by aquaruiz**

### CODE REVIEW

**Comment by aquaruiz** [comment]
@qnislavnachev

File: design-patterns/behavioral-patterns/src/iterator/IcecreamRepository.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+	@Override
+	public Iterator<String> iterator() {
+		Iterator<String> it = new Iterator<String>() {
+
+			private int currentIndex = 0;
+			
+			@Override
+			public boolean hasNext() {
+				return currentIndex < icecreams.length && icecreams[currentIndex] != null;
```
Comment (qnislavnachev) [inline design-patterns/behavioral-patterns/src/iterator/IcecreamRepository.java:38]: You will have a problem in the implementation. if input some ice creams and one of them is 'null' the iteration will not work properly. For Example:
```
new String["vanila", null, "chocolate"]
```

With this array of ice creams the iteration will stop at the second one and the only the first one will be rendered.
You have two options here to not allow null elements in the array or to allow iteration of the null elements

**Review by qnislavnachev — CHANGES_REQUESTED** [review]

File: design-patterns/behavioral-patterns/src/iterator/IcecreamRepository.java (hunk 1 above)
Comment (qnislavnachev) [reply design-patterns/behavioral-patterns/src/iterator/IcecreamRepository.java:38]: The check for null is no longer necessary as we are sure that the array does not contains null elements

**Review by qnislavnachev — CHANGES_REQUESTED** [review]

**Comment by aquaruiz** [comment]
How can we be sure about that?

**Pull request merged**